### ISSUE TITLE
bcbio order of task processing seems to be wasteful concerning storage requirements

### ISSUE THREAD

**Issue by schelhorn** [chapmanb/bcbio-nextgen#799]
Dear Brad & other contributors,

the following is more a request for clarification or a feature request than a bug; however it may be of some interest in others and may also result in code changes, that's why I decided to list it here.

It seems to be as if the order of processing of bcbio currently does not try to limit storage requirements since the same task (e.g., alignment) seems to be scheduled in parallel for all samples rather than only for a subset of samples that can be processed until completion. In other words, if there is a order of tasks n in 1:N for each sample that have to be completed in serial order for completing the analysis of that sample, bcbio seems to schedule task n for completed for _all_ samples before before task n+1 is started, again for all samples. As a result, the maximal storage requirements for a bcbio analysis seem to be quite high, since first all alignments and their post-processed versions have to be stored in the working directory before the result files, for all samples in parallel, are generated. 

In contrast, I suggest it would be more parsimonious to run as many samples until completion as can be processed in parallel (CPU-wise and memory-wise), then upload the results to the final `upload` destination and finally remove all data pertaining to these samples from the working directory (perhaps by leaving behind a lock-file indicating a sample has finished, to allow for a clean restart after errors), before  scheduling tasks for the next set of samples. In this manner, storage requirements would be significantly lessened and scalability improved. I realize that for some analysis steps (e.g., joint variant calling, but there may be others) certain output data have to be preserved well after completion of a sample to allow for across-sample analyses; however, these data could then be retained in the working directory as required and only deleted after they are not longer required.

The reason why I would prefer such a strategy is that high-bandwidth clustered network storage is expensive and at our organization we do not have enough of it for moving a couple of hundred of exomes across it while also storing all alignments in parallel. Many labs have a tiered storage architecture of (and your notation may differ here) tier 1 node-local storage (e.g., raid-0 SSDs), tier 2 clustered network storage (e.g., lustre, or in principle also HDFS), tier 3 non-clustered network storage (e.g., NFS), and tier 4 archive storage (more slow NFS, S3, tape). However, for highly parallel processing on a job scheduler you are basically limited to using tier 2 storage (plus tier 1 for local temporary files) for performance and parallel access reasons.

Please correct me of bcbio already follows such a execution strategy, or offers an option for doing so – my observations are relying on limited experience with a bcbio test instance we plan to move to production eventually, and I may have misinterpreted the data. Either way, an option for providing automatic deletion of intermediary files in the working directory that are not used by other analysis steps and also not uploaded to the final `upload` destination would be valuable to save storage, even if the execution strategy is fine or too much hassle to change.

**Comment by chapmanb** [comment]
Sven-Eric;
Thanks for the thoughts. I'm agreed with your assessment, it would be better to have more fine grained control over removal of intermediates and scheduling of tasks. One of the things we're working on is integration with the common workflow language (https://github.com/common-workflow-language/common-workflow-language) which would allow bcbio to sit on top of other platforms like Arvados, Seven Bridges or Galaxy that handle these kinds of file expirations and management.

In the short term, our normal workaround is to do smaller batches where you don't need to process all the files in parallel. Would this work? I'm not sure if you're calling all these together (in which case we need those BAM files) or are running a lot of independent batches together. In general we try not to have too many extra intermediates around that won't be transferred to upload later but could also look to optimize this some if that would help as a workaround.

Thanks again for the suggestions and thoughts.

**Comment by schelhorn** [comment]
Hi Brad, thanks for the clarification. I still believe that assigning priority to completing analysis for single samples rather than running all samples in parallel would make sense. In that manner we could avoid generating all these large intermediary files (mostly different versions of BAM files) _at the same time_, thus limiting the maximum storage requirements. Still, I realize that would probably mean to tinker with the IPython scheduler, and that may be work-intensive. I'll see of there are large files in the working directory that could be deleted prior to upload, but I guess you guys have already done that. I'll let you know if I find something.

**Comment by chapmanb** [comment]
Sven-Eric;
I absolutely agree it would be nice to be able to custom prioritize sections of the analysis. Practically, bcbio contains a very simple scheduler which executes tasks serially allowing for scatter/gather at each step. We purposely did not want to build a task scheduler but instead focus on the analysis/biological components. The hope is we'll be able to integrate with the common workflow language to take advantage of infrastructure that would enable more complex workflows with graph-based representations, prioritization and intermediate file expiration.

For anything immediate, happy to try to minimize intermediates as much as possible.

Thanks again for the feedback and hope this helps provide more context around the current infrastructure.

**Comment by lpantano** [comment]
Hi 

I am closing this because it seems an old issue. Come back with other suggestion since the pipeline changed a lot since then.

cheers
